core/templates/pages/contributor-dashboard-page/opportunities-list/opportunities-list.component.ts

```diff
@@ -20,17 +20,98 @@ require(
   'pages/contributor-dashboard-page/opportunities-list-item/' +
   'opportunities-list-item.component.ts');
 
+import { Subscription } from 'rxjs';
+
 angular.module('oppia').component('opportunitiesList', {
   bindings: {
-    opportunityDataIsLoading: '=',
-    getOpportunities: '&opportunities',
-    isMoreOpportunitiesAvailable: '&moreAvailable',
-    isLabelRequired: '&labelRequired',
-    isProgressBarRequired: '&progressBarRequired',
-    onLoadMoreOpportunities: '&',
-    onClickActionButton: '=',
-    getOpportunityHeadingTruncationLength:
-      '&opportunityHeadingTruncationLength'
+    loadOpportunities: '<',
+    labelRequired: '<',
+    progressBarRequired: '<',
+    loadMoreOpportunities: '<',
+    onClickActionButton: '<',
+    opportunityHeadingTruncationLength: '<'
   },
-  template: require('./opportunities-list.component.html')
+  template: require('./opportunities-list.component.html'),
+  controller: [
+    '$rootScope', '$scope', 'TranslationLanguageService',
+    'OPPORTUNITIES_PAGE_SIZE', function(
+        $rootScope, $scope, TranslationLanguageService,
+        OPPORTUNITIES_PAGE_SIZE) {
+      var ctrl = this;
+      var opportunities = [];
+      ctrl.visibleOpportunities = [];
+      ctrl.directiveSubscriptions = new Subscription();
+
+      ctrl.directiveSubscriptions.add(
+        TranslationLanguageService.onActiveLanguageChanged.subscribe(
+          () => ctrl.$onInit()));
+
+      if ($scope.$parent.$ctrl.hasOwnProperty(
+        'reloadOpportunitiesEventEmitter')) {
+        ctrl.directiveSubscriptions.add(
+          $scope.$parent.$ctrl.reloadOpportunitiesEventEmitter.subscribe(
+            () => ctrl.$onInit()));
+      }
+
+      if ($scope.$parent.$ctrl.hasOwnProperty(
+        'removeOpportunityEventEmitter')) {
+        ctrl.directiveSubscriptions.add(
+          $scope.$parent.$ctrl.removeOpportunityEventEmitter.subscribe(
+            (opportunityId) => {
+              opportunities = opportunities.filter(function(opportunity) {
+                return opportunity.id !== opportunityId;
+              });
+              ctrl.visibleOpportunities = opportunities.slice(
+                0, OPPORTUNITIES_PAGE_SIZE);
+            }));
+      }
+
+      ctrl.$onDestroy = function() {
+        ctrl.directiveSubscriptions.unsubscribe();
+      };
+
+      ctrl.$onInit = function() {
+        opportunities = [];
+        ctrl.visibleOpportunities = [];
+        ctrl.activePageNumber = 1;
+        ctrl.lastPageNumber = 1000;
+        ctrl.loadingOpportunityData = true;
+
+        ctrl.loadOpportunities().then(function({opportunitiesDicts, more}) {
+          opportunities = opportunitiesDicts;
+          ctrl.visibleOpportunities = opportunities.slice(
+            0, OPPORTUNITIES_PAGE_SIZE);
+          ctrl.lastPageNumber = more ? ctrl.lastPageNumber : Math.ceil(
+            opportunities.length / OPPORTUNITIES_PAGE_SIZE);
```

Comment:
Maybe extract this into a function, as it's duplicated below as well?

Reply:
It's linear and simple and only used twice so I left it as-is for now.